### ISSUE TITLE
Setting maxEntrySize does not truncate big json payloads correctly

### ISSUE THREAD

**Issue by leononame** [googleapis/nodejs-logging#1087]
When setting maxEntrySize for a logger and using `jsonPayload`, the truncation algorithm does not truncate all fields, but rather only a very small set of fields: https://github.com/googleapis/nodejs-logging/blob/master/src/log.ts#L984

As a user, when passing maxEntrySize, I expect it to also truncate custom fields. In our case, one of our fields is a huge list of UUIDs that does not matter if it gets truncated a little, but should be there. However, it does not get truncated and in the end we get exceptions thrown in prod because our log entries are too big.

**Comment by minherz** [comment]
Hi @leononame,
thank you for submitting your inquiry. I understand that your would like to control which fields in the submitted log payload will be truncated based on the total payload cap.
Would you mind to elaborate more about the customary logic you are proposing.
* What should be an order of the truncated fields?
* If there are several truncated fields, what should the truncating logic for each one of them: truncate to zero length, up to predefined min length that is defined per field or common?
* Should `maxEntrySize` be enforced if the total size of the log entry goes beyond the [limit](https://cloud.google.com/logging/quotas#log-limits)?

Thank you

**Comment by leononame** [comment]
Hi @minherz 
I guess the least I would expect is that when setting maxEntrySize, the message actually gets truncated to that size even when it's a JSON payload. The current implementation, linked in the issue, does not guarantee that at all. It just blindly truncates some fields and never checks again for message size. This leads to exceptions in production when logging objects that are too big. 

I think there are a couple of subtleties that should be changed
1. Currently, on a JSON payload the following fields get always truncated:
  ```ts
        const fieldsToTruncate = [
          // Winston:
          'jsonPayload.fields.metadata.structValue.fields.stack.stringValue',
          // Bunyan:
          'jsonPayload.fields.msg.stringValue',
          'jsonPayload.fields.err.structValue.fields.stack.stringValue',
          'jsonPayload.fields.err.structValue.fields.message.stringValue',
          // All:
          'jsonPayload.fields.message.stringValue',
        ];
  ```
  Even if only truncating the first field would suffice to make the payload small enough. I think it'd be beneficial if not all fields get truncated blindly.
2. I would also expect the function to truncate additional fields if the message is still too big (i.e., to truncate fields passed by the user). It should probably start truncating the biggest fields first so that it has to remove as few fields as needed.
3. Ideally, during setup one could specify a list of fields that should be prioritized for truncating, but I don't see this as an essential addition.

**Comment by leononame** [comment]
I see that 1. and 3. have been addressed by PR #1177, but for 2., there's still something open: The implementation will still throw errors when logging big objects even though maxEntrySize is set, because it does not truncate fields not specified in the truncate list. If you want to keep that behavior, I'd ask you to at least document it.

**Comment by losalex** [comment]
Thanks @leononame for your response!
Indeed, the logic for truncating entries is not ideal and it is hard to build generic solution which could address every specific scenario. 
I have another question - did you tried to use [partialSuccess](https://github.com/googleapis/nodejs-logging/blob/51d6efd3341e6a02ffd130e36eb1491f410e1dc1/src/log.ts#L35)? Based on explanation [here](https://cloud.google.com/logging/docs/reference/v2/rest/v2/entries/write#request-body) it should drop only oversized entries and still log others.
